pmlc/dialect/tile/ir/ops.td

```diff
@@ -217,3 +219,87 @@ def PragmaOp : MaterializeOp<"pragma", [
 
   let assemblyFormat = "$tensor $op $attrs attr-dict `:` type($tensor)";
 }
+
+class LoopBaseOp<string mnemonic, list<OpTrait> traits = []> :
+    Op<TileDialect, mnemonic, traits> {
+  // For every standard op, there needs to be a:
+  //   * void print(OpAsmPrinter &p, ${C++ class of Op} op)
+  //   * LogicalResult verify(${C++ class of Op} op)
+  //   * ParseResult parse${C++ class of Op}(OpAsmParser &parser,
+  //                                         OperationState &result)
+  // functions.
+  let parser = [{ return pmlc::dialect::tile::parse$cppClass(parser, result); }];
+  let printer = [{ return pmlc::dialect::tile::print(p, *this); }];
+  let verifier = [{ return pmlc::dialect::tile::verify(*this); }];
+}
+
+def LoopOp : LoopBaseOp<"loop",
```

Comment:
Yes, in a earlier version we directly created `scf.for` for edsl Loop Op at tile level. However, later we found that to implement the OV Loop op , we will need to also add `scf.while` (and maybe other scf ops at tile level), a implementation of other features `loop break` will be needed at tile level, to support parameters like `upperbound`(or maxTripCount in OV Loop) as a 1D tensor(instead of a constant tensor), we will also need to add load&store of Tensor at tile level. So we thought maybe it is better to implement a Loop op at tile level, all these features can be implemented later at tile to pxa level.

Reply:
Before we go any further, let's look at it from another angle. What primitives do we need to have to support more OV ops? I think before we decide on a design, let's write the psuedocode for these ops as if we had new primitives in place. This will help constrain the design and let us review exactly what's necessary from any new primitives.

Reply:
Thanks, psuedocode for these ops that needs loop would be very helpful, we will work on this.

Reply:
So, I also would prefer to use the scf ops, although clearly it does add more work (and requires bringing in tensor.extract etc).  Personally, I think option #1 above is best.  My reasoning is basically:  If we go with option 2 or 3, we are pushing OpenVino semantics down into the EDSL and PlaidML stack respectively.  This will make integration with OpenVino easier, but it will make integration with other frameworks more difficult, or at least less natural.  Long term, we intend to try to move much of PlaidML upstream, so whenever we can prefer matching upstream design, I think we should.

Reply:
Thanks. Maybe in this PR I should focus on the basic for-loop mode (remove tile.LoopOp and tileYieldOp and replace edsl loopOp by ForLoopOp and convert to scf.for at tile level). While we write psuedocode or implement other OV ops that depends on loop, we can add follow-up PRs that add needed features. What do you think? @jbruestle @flaub @tzerrell

Reply:
FYI. 
Current edsl loop (or for_loop) op in this PR should be able to replace all c++ for loops used in many ops implemented in the OV plugin. Using it can avoid generating repeated tile code.

However, for quite a few remaining ops that currently blocked by loop (and usually also blocked by dynamic shape issue), the c++ for loop or edsl for_loop in this PR are current not sufficient.
 
Below is a simplified common pattern involving loop and dynamic shape that we needs in a lot of ops but is not currently supported.

```
auto IX = Placeholder(DType::INT32, {1});
Tensor zero = edsl::reshape(edsl::index({TensorDim(1)}, 0), {1, 1});

TensorVec O = for_loop(IX, {zero, zero}, [&](TensorVec iv) {
  iv[1] = op::concatenate({iv[1], iv[0]}, 1);
  iv[0] = iv[0] + 1;
  return iv;
};)

O[1] is the tensor we want to use in following code
...

```
where the `IX` is a 1D tensor with shape `{1}`, which can be either a `Placeholder`  or a result from other edsl ops. We need to pass it as the `upperbound` of the for_loop (this is currently supported in this PR). The value (let's say `IX_val`) in `Tensor IX` is not known to us when we compile the tile program, but we want to get a 1D tensor (out of from the loop) with size = `IX_val`, and values from 0,1,2,... to `IX_val` - 1. However this changes the loop iterator tensor shape in each iteration, so it is not supported now.

Now we are working on resolving this issue as well. This issue mainly related to dynamic shape issue, so if we have a fix, it would probably be better to be submited in a separate PR.

Once this common pattern can be supported, we should be able to implement most of the ops currently blocked by loop and dynamic shape (Separately, the while loop, infinite loop or other features can be added later to implement OV Loop and related Ops).  

@jbruestle  @flaub @tzerrell  Any comments or suggestions are much appreciated. Thanks.